dns_server/src/crawler_p2p/crawler/address_data.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use utils::const_value::ConstValue;
+
+/// Address state transition
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AddressTransition {
+    Connecting,
+    Connected,
+    Disconnecting,
+    Disconnected,
+}
+
+/// When the server drops the unreachable node address. Used for negative caching.
+pub const PURGE_UNREACHABLE_TIME: Duration = Duration::from_secs(3600);
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+pub const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AddressState {
+    Connecting {
+        /// The number of consecutive failed connection attempts.
+        /// New connection attempts are made after a progressive backoff time.
+        fail_count: u32,
+
+        /// Whether the address was reachable at least once.
+        /// Addresses that were once reachable are stored in the DB.
+        was_reachable: bool,
+    },
+
+    Connected {},
+
+    Disconnecting {
+        /// Same as above
+        fail_count: u32,
+
+        /// Same as above
+        was_reachable: bool,
+    },
+
+    Disconnected {
+        /// Same as above
+        fail_count: u32,
+
+        /// Same as above
+        was_reachable: bool,
+
+        /// The time when the address went into the disconnected state
+        disconnected_at: Duration,
+    },
+
+    PermanentlyUnreachable {
```

Comment:
If I understand correctly, perhaps you mean something like "never reached so far". `PermanentlyUnreachable` would mean that it's unreachable forever by design. I think we can use something `NeverReachedYet`.

Reply:
This is a final state where an address is marked as unreachable and there will be no more attempts to connect to it. After `erase_after` time it would be removed from memory (it can be added as new after that).
I will rename the state to `Unreachable` and add a comment.

Reply:
Cool.